hd44780/hd44780.go

```diff
@@ -14,216 +14,427 @@ import (
 	"time"
 
 	"periph.io/x/conn/v3"
+	"periph.io/x/conn/v3/display"
 	"periph.io/x/conn/v3/gpio"
 )
 
-// lineTwo offset for the second line in the LCD buffer.
-const lineTwo = 0x40
+type writeMode bool
 
-// Dev is the 4-bit addressing device for HD-44780
-type Dev struct {
-	// data pins
-	dataPins []gpio.PinOut
+type ifMode byte
 
-	// register select pin
-	rsPin gpio.PinOut
+const (
+	modeCommand writeMode = false
+	modeData    writeMode = true
 
-	// enable pin
+	cmdByte byte = 0xfe
+
+	mode4Bit ifMode = 0x04
+	mode8Bit ifMode = 0x08
+)
+
+// HD44780 is an implementation that supports writing to LCD displays using a
+// gpio.Group for the data pins, and discrete pins for the reset, enable, and
+// and backlight pins.
+//
+// Implements periph.io/conn/x/display/TextDisplay and display.DisplayBacklight
+type HD44780 struct {
+	dataPins  gpio.Group
+	resetPin  gpio.PinOut
 	enablePin gpio.PinOut
+	blMono    display.DisplayBacklight
+	blRGB     display.DisplayRGBBacklight
+	mode      ifMode
+	rows      int
+	cols      int
+	on        bool
+	cursor    bool
+	blink     bool
+	lastWrite int64
 }
 
-// New creates and initializes the LCD device
-//
-//	data - references to data pins
-//	rs - rs pin
-//	e - strobe pin
-func New(data []gpio.PinOut, rs, e gpio.PinOut) (*Dev, error) {
-	if len(data) != 4 {
-		return nil, fmt.Errorf("expected 4 data pins, passed %d", len(data))
-	}
-	dev := &Dev{
-		dataPins:  data,
-		enablePin: e,
-		rsPin:     rs,
-	}
-	if err := dev.Reset(); err != nil {
-		return nil, err
+const (
+	delayCommand   time.Duration = 2000
+	delayCharacter time.Duration = 200
+)
+
+var rowConstants = [][]byte{{0, 0, 64}, {0, 0, 64, 20, 84}}
+var clearScreen = []byte{cmdByte, 0x01}
+var goHome = []byte{cmdByte, 0x02}
+var setCursorPosition = []byte{cmdByte, 0x80}
+
+// Return the row offset value
+func getRowConstant(row, maxcols int) byte {
+	var offset int
+	if maxcols != 16 {
+		offset = 1
 	}
-	return dev, nil
+	return rowConstants[offset][row]
 }
 
-// Reset resets the HC-44780 chipset, clears the screen buffer and moves cursor to the
-// home of screen (line 0, column 0).
-func (r *Dev) Reset() error {
-	if err := r.clearBits(); err != nil {
-		return err
-	}
+// NewHD44780 takes a GPIO group, and gpio.PinOut for reset and enable. It
+// returns an HD44780 device in an initialized state and ready for use.
+//
+// The first 4 or 8 pins of the data group must be connected to the data lines
+// To use 4 bit mode, you would connect lines D4-D7 on the display, and for 8
+// bit mode, D0-D7. If dataPinGroup is 8 or more pins, then it's assumed the
+// display is connected using all 8 pins.
+//
+// backlight should implement either display.DisplayBacklight or
+// display.DisplayRGBBacklight. See GPIOMonoBacklight.
+func NewHD44780(
+	dataPinGroup gpio.Group,
+	resetPin, enablePin *gpio.PinOut,
```

Comment:
gpio.PinOut is an interface, so you do not want to take the pointer, it is already a pointer.